File: app/src/main/java/br/com/jonathanzanella/myexpenses/views/MainActivity.java
package br.com.jonathanzanella.myexpenses.views;

import android.app.SearchManager;
import android.content.Intent;
import android.os.Bundle;
import android.support.design.widget.NavigationView;
import android.support.design.widget.TabLayout;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v7.widget.SearchView;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;

import br.com.jonathanzanella.myexpenses.R;
import br.com.jonathanzanella.myexpenses.account.AccountView;
import br.com.jonathanzanella.myexpenses.bill.BillView;
import br.com.jonathanzanella.myexpenses.card.CardView;
import br.com.jonathanzanella.myexpenses.expense.ExpenseView;
import br.com.jonathanzanella.myexpenses.log.LogsView;
import br.com.jonathanzanella.myexpenses.overview.OverviewExpensesView;
import br.com.jonathanzanella.myexpenses.receipt.ReceiptView;
import br.com.jonathanzanella.myexpenses.resume.ResumeView;
import br.com.jonathanzanella.myexpenses.source.SourceView;
import br.com.jonathanzanella.myexpenses.sync.SyncView;
import butterknife.Bind;

/**
 * Created by Jonathan Zanella on 25/01/16.
 */
public class MainActivity extends BaseActivity implements NavigationView.OnNavigationItemSelectedListener, SearchView.OnQueryTextListener {
	@Bind(R.id.act_main_drawer)
	DrawerLayout drawer;
	@Bind(R.id.act_main_navigation_view)
	NavigationView navigationView;
	@Bind(R.id.act_main_content)
	FrameLayout content;
	@Bind(R.id.tabs)
	TabLayout tabs;

	private BaseView currentView;
	private String filter;

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_main);
	}

	@Override
	protected void onPostCreate(Bundle savedInstanceState) {
		super.onPostCreate(savedInstanceState);

		ActionBarDrawerToggle drawerToggle = new ActionBarDrawerToggle(
				this, drawer, toolbar,
				R.string.navigation_drawer_open, R.string.navigation_drawer_close
		);

		navigationView.setNavigationItemSelectedListener(this);
		drawer.setDrawerListener(drawerToggle);
		drawerToggle.syncState();

		addViewToContent(new ResumeView(this));
	}

	@Override
	protected void onActivityResult(int requestCode, int resultCode, Intent data) {
		super.onActivityResult(requestCode, resultCode, data);

		for(int i = 0; i < content.getChildCount(); i++) {
			View v = content.getChildAt(i);
			if(v instanceof BaseView)
				((BaseView)v).onActivityResult(requestCode, resultCode, data);
		}
	}

	@Override
	public boolean onCreateOptionsMenu(Menu menu) {
		MenuInflater inflater = getMenuInflater();
		inflater.inflate(R.menu.options_menu, menu);

		SearchManager searchManager = (SearchManager) getSystemService(SEARCH_SERVICE);
		SearchView searchView = (SearchView) menu.findItem(R.id.search).getActionView();
		searchView.setSearchableInfo(searchManager.getSearchableInfo(getComponentName()));
		searchView.setOnQueryTextListener(this);

		return true;
	}

	@Override
	public boolean onQueryTextSubmit(String query) {
		return false;
	}

	@Override
	public boolean onQueryTextChange(String newText) {
		filter = newText;
		currentView.filter(newText);
		return false;
	}

	@Override
	protected void onResume() {
		super.onResume();

		for(int i = 0; i < content.getChildCount(); i++) {
			View v = content.getChildAt(i);
			if(v instanceof BaseView)
				((BaseView)v).refreshData();
		}
	}

	private void addViewToContent(BaseView child) {
		content.removeAllViews();
		child.setLayoutParams(new ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT,
				ViewGroup.LayoutParams.MATCH_PARENT));
		content.addView(child);
		child.setTabs(tabs);
		child.filter(filter);
		currentView = child;
	}

	@Override
	public boolean onNavigationItemSelected(MenuItem item) {
		switch (item.getItemId()) {
			case R.id.menu_resume:
				addViewToContent(new ResumeView(this));
				setTitle(R.string.app_name);
				drawer.closeDrawers();
				return true;
			case R.id.menu_accounts:
				addViewToContent(new AccountView(this));
				setTitle(R.string.accounts);
				drawer.closeDrawers();
				return true;
			case R.id.menu_sources:
				addViewToContent(new SourceView(this));
				setTitle(R.string.sources);
				drawer.closeDrawers();
				return true;
			case R.id.menu_credit_card:
				addViewToContent(new CardView(this));
				setTitle(R.string.credit_card);
				drawer.closeDrawers();
				return true;
			case R.id.menu_receipts:
				addViewToContent(new ReceiptView(this));
				setTitle(R.string.receipts);
				drawer.closeDrawers();
				return true;
			case R.id.menu_expenses:
				addViewToContent(new ExpenseView(this));
				setTitle(R.string.expenses);
				drawer.closeDrawers();
				return true;
			case R.id.menu_bills:
				addViewToContent(new BillView(this));
				setTitle(R.string.bills);
				drawer.closeDrawers();
				return true;
			case R.id.menu_overview:
				addViewToContent(new OverviewExpensesView(this));
				setTitle(R.string.overview);
				drawer.closeDrawers();
				return true;
			case R.id.menu_sync:
				addViewToContent(new SyncView(this));
				setTitle(R.string.sync);
				drawer.closeDrawers();
				return true;
			case R.id.menu_log:
				addViewToContent(new LogsView(this));
				setTitle(R.string.log);
				drawer.closeDrawers();
				return true;
		}
		return false;
	}
}

File: app/src/main/java/br/com/jonathanzanella/myexpenses/views/MainActivity.kt
package br.com.jonathanzanella.myexpenses.views

import android.app.SearchManager
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.support.annotation.IdRes
import android.support.design.widget.NavigationView
import android.support.v7.app.ActionBarDrawerToggle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.SearchView
import android.support.v7.widget.Toolbar
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import br.com.jonathanzanella.myexpenses.R
import br.com.jonathanzanella.myexpenses.account.AccountView
import br.com.jonathanzanella.myexpenses.bill.BillView
import br.com.jonathanzanella.myexpenses.card.CardView
import br.com.jonathanzanella.myexpenses.expense.ExpenseView
import br.com.jonathanzanella.myexpenses.log.LogsView
import br.com.jonathanzanella.myexpenses.overview.OverviewExpensesView
import br.com.jonathanzanella.myexpenses.receipt.ReceiptView
import br.com.jonathanzanella.myexpenses.resume.ResumeView
import br.com.jonathanzanella.myexpenses.source.SourceView
import br.com.jonathanzanella.myexpenses.sync.SyncView
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity(), NavigationView.OnNavigationItemSelectedListener, SearchView.OnQueryTextListener {
    private lateinit var currentView: View
    private var filter = ""
    private var selectedItem = -1

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        storeBundle(savedInstanceState)
        storeBundle(intent.extras)
    }

    override fun onPostCreate(savedInstanceState: Bundle?) {
        super.onPostCreate(savedInstanceState)
        setSupportActionBar(toolbar as Toolbar?)

        val drawerToggle = ActionBarDrawerToggle(
                this, drawer, toolbar as Toolbar?,
                R.string.navigation_drawer_open, R.string.navigation_drawer_close
        )

        navigationView!!.setNavigationItemSelectedListener(this)

        @Suppress("DEPRECATION")
        drawer!!.setDrawerListener(drawerToggle)
        drawerToggle.syncState()

        if (selectedItem == -1)
            selectMenuItem(R.id.menu_resume)
    }

    fun storeBundle(extras: Bundle?) {
        if (extras != null && extras.containsKey(KEY_SELECTED_ITEM))
            selectMenuItem(extras.getInt(KEY_SELECTED_ITEM))
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putInt(KEY_SELECTED_ITEM, selectedItem)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)

        (0..content!!.childCount - 1)
                .map { content!!.getChildAt(it) }
                .forEach { (it as? BaseView)?.onActivityResult(requestCode, resultCode, data) }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.options_menu, menu)

        val searchManager = getSystemService(Context.SEARCH_SERVICE) as SearchManager
        val searchView = menu.findItem(R.id.search).actionView as SearchView
        searchView.setSearchableInfo(searchManager.getSearchableInfo(componentName))
        searchView.setOnQueryTextListener(this)

        return true
    }

    override fun onQueryTextSubmit(query: String): Boolean {
        return false
    }

    override fun onQueryTextChange(newText: String): Boolean {
        filter = newText
        val view = currentView
        if(view is BaseView)
            view.filter(newText)
        return false
    }

    override fun onResume() {
        super.onResume()

        (0..content.childCount - 1)
                .map { content.getChildAt(it) }
                .filterIsInstance<RefreshableView>()
                .forEach { it.refreshData() }
    }

    private fun addViewToContent(child: View) {
        content.removeAllViews()
        child.layoutParams = ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.MATCH_PARENT)
        content!!.addView(child)
        if(child is BaseView) {
            child.setTabs(tabs)
            child.filter(filter)
        }
        currentView = child
    }

    override fun onNavigationItemSelected(item: MenuItem): Boolean {
        return selectMenuItem(item.itemId)
    }

    private fun selectMenuItem(@IdRes itemId: Int): Boolean {
        selectedItem = itemId
        navigationView!!.setCheckedItem(selectedItem)
        when (itemId) {
            R.id.menu_resume -> {
                addViewToContent(ResumeView(this))
                setTitle(R.string.app_name)
                drawer!!.closeDrawers()
                return true
            }
            R.id.menu_accounts -> {
                addViewToContent(AccountView(this))
                setTitle(R.string.accounts)
                drawer!!.closeDrawers()
                return true
            }
            R.id.menu_sources -> {
                addViewToContent(SourceView(this))
                setTitle(R.string.sources)
                drawer!!.closeDrawers()
                return true
            }
            R.id.menu_credit_card -> {
                addViewToContent(CardView(this))
                setTitle(R.string.credit_card)
                drawer!!.closeDrawers()
                return true
            }
            R.id.menu_receipts -> {
                addViewToContent(ReceiptView(this))
                setTitle(R.string.receipts)
                drawer!!.closeDrawers()
                return true
            }
            R.id.menu_expenses -> {
                addViewToContent(ExpenseView(this))
                setTitle(R.string.expenses)
                drawer!!.closeDrawers()
                return true
            }
            R.id.menu_bills -> {
                addViewToContent(BillView(this))
                setTitle(R.string.bills)
                drawer!!.closeDrawers()
                return true
            }
            R.id.menu_overview -> {
                addViewToContent(OverviewExpensesView(this))
                setTitle(R.string.overview)
                drawer!!.closeDrawers()
                return true
            }
            R.id.menu_sync -> {
                addViewToContent(SyncView(this))
                setTitle(R.string.sync)
                drawer!!.closeDrawers()
                return true
            }
            R.id.menu_log -> {
                addViewToContent(LogsView(this))
                setTitle(R.string.log)
                drawer!!.closeDrawers()
                return true
            }
        }
        selectedItem = -1
        navigationView!!.setCheckedItem(selectedItem)
        return false
    }

    companion object {
        private val KEY_SELECTED_ITEM = "KeySelectedItem"
    }
}
